src/stream/transforms.ts

```diff
@@ -291,4 +291,96 @@ export class StreamTransforms<T, E> extends StreamConsumption<T, E> {
             }
         });
     }
+
+    /**
+     * Map over each value in the stream, and apply some callback to it. This differs from `map` as
+     * the upstream is continually pulled, regardless if the downstream is ready for it and whether
+     * the current iteration has complete.
+     *
+     * # Example
+     *
+     * Approximate timelines are shown below. It isn't 100% correct for what's actually happening,
+     * but gives the right idea.
+     *
+     * - `p`: producer (upstream)
+     * - `m`: map operation (asynchronous)
+     *
+     * ## Map
+     *
+     * ```
+     *  0ms       10ms      20ms      30ms
+     *  |---------|---------|---------|---------|
+     *  |<-- p -->|<-- m -->|         |         |
+     *  |         |         |<-- p -->|<-- m -->|
+     * ```
+     *
+     * ## Buffered Map
+     *
+     * ```
+     *  0ms       10ms      20ms      30ms
+     *  |---------|---------|---------|---------|
+     *  |<-- p -->|<-- m -->|         |         |
+     *  |         |<-- p -->|<-- m -->|         |
+     *  |         |         |<-- p -->|<-- m -->|
+     * ```
+     *
+     * @group Transform
+     */
+    bufferedMap<U>(
+        cb: (value: T) => MaybePromise<MaybeAtom<U, E>>,
+        options?: { delay?: number },
+    ): Stream<U, E> {
+        const trace = this.trace("bufferedMap");
+
+        let signal = newSignal();
+        let end = false;
+
+        // Buffer all of the pending map results
+        const buffer: Promise<Atom<U, E>>[] = [];
+
+        // Continually fill up the buffer
+        (async () => {
+            // Continually pull items from the stream
+            for await (const atom of this) {
+                if (isOk(atom)) {
+                    // Pass the value through the callback
+                    buffer.push(handler(() => cb(atom.value), trace));
+                } else {
+                    // Add the atom directly to the buffer
+                    buffer.push(Promise.resolve(atom));
```

Comment:
I wonder if we should also have an option for setting a `maxBufferSize` – imagine a scenario where your downstream consumer has either stalled or is incredibly slow, this function will continuously buffer up items until it consumes all memory.

Another alternative could be to have a `maxBufferPeriod` to only buffer for say 30s before waiting for the consumer to catch up. #recommended

Reply:
I'll add both of these as options. I do currently have a `delay` option so that the upstream isn't overwhelmed, but these are good options too.

How should it treat these by default? Should some/all of the options have some default value, or allow it to be unbounded if nothing is supplied?

Reply:
It's a good question. I think a time-based `maxBufferPeriod` feels too arbitrary because it's hard to pre-determine how fast the upstream could be producing so I don't think we'd want a default there.

A default `maxBufferSize` makes sense and is common practice across a lot of systems to save people who forget to set one. Maybe something sensible like `8000`? Should be small enough to prevent crashes (hard to know for sure without knowing each atom size) but large enough to be useful. Then allow people to increase it further if necessary.